[TensorPipe] Implement join correctly

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* **#38933 [TensorPipe] Implement join correctly**
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging



Based on what I could understand from how the RPC shutdown operates and from what the ProcessGroup agent does, the join method is supposed to act as a barrier among all workers that waits until they all have finished all their pending work, including work that may be triggered by nested calls or by callbacks.


Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020/)

Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020)

lw committed:
[TensorPipe] Implement join correctly

My understanding of how the RPC shutdown works is as follows:
- All workers enter a barrier (i.e., `_wait_all_workers`) so that "passive" workers (the ones that don't initiate any RPC call themselves, but just receive and handle requests from other workers) stay alive until the "active" workers have also completed. This way we know that when the barrier returns all workers are ready to shut down.
- Then we start shutting down the layers of the stack that are between the user and the "core" RPC agent, namely the dist autograd context and the rref context. These cleanups trigger some asynchronous RPC calls but, apparently, do *not* wait for their futures to complete. (This is because if they did then this would also cause some operations to become blocking during normal operations).
- Then we call `join` on the RPC worker (more on this later).
- Then we do another barrier.
- Then we shutdown the agent (abruptly, aborting any pending operation).

From the above, I believe the only requirement that is put on `join` in order for the shutdown to be graceful (i.e., for no operation to actually be aborted) is for `join` to wait for all *outgoing* requests to finish or, equivalently, for all pending futures to be completed. In other words, the "client" part of the agent needs to flush. I don't think we need to worry about the "server" part of the agent, i.e., we do not need to worry about waiting for the requests the agent is currently executing on behalf of other workers, because we will implicitly wait for them in the barrier that follows the `join`.

Based on this understanding, in this diff I am changing the `join` method of the TensorPipe agent to wait until `clientActiveCalls_` reaches zero.

Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Implement join correctly"




Based on what I could understand from how the RPC shutdown operates and from what the ProcessGroup agent does, the join method is supposed to act as a barrier among all workers that waits until they all have finished all their pending work, including work that may be triggered by nested calls or by callbacks.


Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020/)

Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -472,11 +472,13 @@ PyObject* rpc_init(PyObject* /* unused */) {
               std::string /* selfName */,
               worker_id_t /* selfId */,
               int /* worldSize */,
+              std::shared_ptr<::c10d::ProcessGroup> /* processGroup */,
```
Is this API what we plan to release in v1.6?

Since we don't need this until shutdown, I wonder if we can create this process group in `join` internally instead of asking applications to pass this in. In the args, we already have `rank` and `world_size`, can we let `rank=0` figure out a port, and send that port to all other peers, then they all use that as the `init_method` to create the process group instance.

lw replied:
Personally I didn't consider this to be part of the API. Final users are supposed to enable the TensorPipe agent by calling:
```
rpc.init_rpc(name="foo", rank=0, world_size=2, backend=rpc.backend_registry.BackendType.TENSORPIPE, rpc_backend_options=rpc.TensorPipeRpcBackendOptions(...))
```
So I consider the API to consist of:
- the "name" `TENSORPIPE`,
- the arguments of `TensorPipeRpcBackendOptions`.

The constructor of the TensorPipeAgent is only invoked by the `_tensorpipe_init_backend_handler` and I considered it to be sort-of a "private" function.

Does this make sense? Or is the constructor of the agent a public API?

lw replied:
As for constructing the process group on-demand, we could definitely do that, I just thought it was easier to construct it in Python (by copying the code used to construct it for the ProcessGroup agent) and then pass it to the agent. I can do as you prefer, but if you want me to construct it in C++ you'll have to show me how to do so. :)

mrshenli replied:
I see. This is good enough then. Let's land as-is.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -667,16 +701,33 @@ std::string TensorPipeAgent::getDefaultIPAddress() {
+void TensorPipeAgent::increaseCallCount(int32_t& count) {
+  std::unique_lock<std::mutex> lock(callCountMutex_);
+  ++count;
+  callCountCV_.notify_all();
```
Do we need to unlock before notify all?

lw replied:
There is no _need_ to unlock before notifying, but you're right that there's also no need to _hold_ the lock while notifying. In fact, in some cases it's more efficient to release the lock before notifying, as otherwise the thread that gets woken up would just immediately block again on acquiring the lock. However, pthreads optimizes for that and there's no performance penalty. The docs for `notify_one` has an entire paragraph dedicated to this: https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

In the end, I'm going to release the lock, it's cleaner.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -667,16 +701,33 @@ std::string TensorPipeAgent::getDefaultIPAddress() {
+void TensorPipeAgent::decreaseCallCount(int32_t& count) {
+  std::unique_lock<std::mutex> lock(callCountMutex_);
+  --count;
+  callCountCV_.notify_all();
```
ditto

lw committed:
Update on "[TensorPipe] Implement join correctly"




Based on what I could understand from how the RPC shutdown operates and from what the ProcessGroup agent does, the join method is supposed to act as a barrier among all workers that waits until they all have finished all their pending work, including work that may be triggered by nested calls or by callbacks.


Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020/)

Differential Revision: [D21703020](https://our.internmc.facebook.com/intern/diff/D21703020)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -171,21 +171,58 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+
+    process_group_timeout = rpc_constants.DEFAULT_PROCESS_GROUP_TIMEOUT
+
+    dist.init_process_group(
```
Let's land this one for now, and add an issue (similar to #33583) to avoid write the default global pg.

lw replied:
I created #39138 to track this
